libgadget/forcetree.c

```diff
@@ -48,13 +48,13 @@ static void
 force_treeupdate_pseudos(int no, const ForceTree * tree);
 
 static void
-force_create_node_for_topnode(int no, int topnode, struct NODE * Nodes, int bits, int x, int y, int z, int *nextfree, const int lastnode);
+force_create_node_for_topnode(int no, int topnode, struct NODE * Nodes, const Domain * domain, int bits, int x, int y, int z, int *nextfree, const int lastnode);
```

Comment:
If the treebuilder object remembers a pointer to the domain object, then we may be able to get rid of a lot of `Domain`s here.

Reply:
Whether if it is sane to the pointer or not depends on if domain object is used in the entire life cycle of the tree or only transiently during the build stage. It smells like it is the full life time -- since we also need the domain object to export ghosts and manage pseudo particles in treewalk.

Reply:
So the TopLeaves are used in the entire tree life cycle for that purpose, and I store a pointer to them in the Tree. But the rest of the domain (the TopNodes and the Task data) is only used during build. This function in particular is really using the TopNodes. It might be possible to remove a Domain from the forcetree functions that build pseudodata if we wanted, but I left it in so I didn't at some future point accidentally introduce bugs related to the tree->TopLeaves pointer not being initialised yet.